Src/ILGPU/IndexTypes.tt

```diff
@@ -362,8 +362,13 @@ namespace ILGPU
         /// Returns the hash code of this index.
         /// </summary>
         /// <returns>The hash code of this index.</returns>
-        public readonly override int GetHashCode() =>
-            <#= def.Expression(" ^ ", p => $"{p}.GetHashCode()") #>;
+        public readonly override int GetHashCode()
+        {
+            int hash = <#= def.Expression(
+                $"{Environment.NewLine}            hash = (hash << 1) ^ (hash >> 1) ^ ",
+                p => $"{p}.GetHashCode();") #>
+            return hash;
+        }
```

Comment:
SUGGESTION: What about using a ValueTuple? i.e. `(X, Y, Z).ToHashCode()`.
Or will this not work on the GPU because it is using `System.HashCode`?

```CSharp
<# if (dimension == 1) { #>
            <#= def.Expression(", ", p => $"{p}") #>.GetHashCode();
<# } else { #>
            (<#= def.Expression(", ", p => $"{p}") #>).GetHashCode();
<# } #>
```

Reply:
I couldn't find anything about how `ValueTuple.GetHashCode()` is implemented beyond it being guaranteed that equal values will have equal hash codes, so I have to assume it _won't_ solve the issue.

No idea about what'll go on inside the GPU, my concern was solely how it's used on the CPU. I've got a use-case where I've got a dictionary that's keyed on 2D grid locations inside an algorithm that heavily utilizes the GPU, and I didn't want to write a redundant implementation of basically just (int x, int y) for it.

Reply:
Internally, `ValueTuple` it is using `System.HashCode`:
https://github.com/dotnet/runtime/blob/v5.0.7/src/libraries/System.Private.CoreLib/src/System/ValueTuple.cs#L566

Based on the source code, `System.HashCode` is using xxHash32:
https://github.com/dotnet/runtime/blob/v5.0.7/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L51

https://github.com/Cyan4973/xxHash

> xxHash is an Extremely fast Hash algorithm, running at RAM speed limits. It successfully completes the SMHasher test suite which evaluates collision, dispersion and randomness qualities of hash functions. Code is highly portable, and hashes are identical across all platforms (little / big endian).

Reply:
Okay, looking deeper into that, it seems like using the ValueTuple might be a better approach. (Something about spreading the hashcodes out more...)